hazelcast/src/test/java/com/hazelcast/test/compatibility/SamplingSerializationService.java

```diff
@@ -0,0 +1,234 @@
+        delegate.writeObject(out, obj);
+    }
+
+    @Override
+    public <T> T readObject(ObjectDataInput in) {
+        return (T) delegate.readObject(in);
+    }
+
+    @Override
+    public <T> T readObject(ObjectDataInput in, Class aClass) {
+        return (T) delegate.readObject(in, aClass);
+    }
+
+    @Override
+    public <T> T toObject(Object data) {
+        return (T) delegate.toObject(data);
+    }
+
+    @Override
+    public <T> T toObject(Object data, Class klazz) {
+        return (T) delegate.toObject(data, klazz);
+    }
+
+    @Override
+    public byte[] toBytes(Object obj) {
+        byte[] bytes = delegate.toBytes(obj);
+        sampleObject(obj, bytes);
+        return bytes;
+    }
+
+    @Override
+    public byte[] toBytes(Object obj, int leftPadding, boolean insertPartitionHash) {
+        byte[] bytes =  delegate.toBytes(obj, leftPadding, insertPartitionHash);
+        sampleObject(obj, bytes);
+        return bytes;
+    }
+
+    @Override
+    public ClassLoader getClassLoader() {
+        return delegate.getClassLoader();
+    }
+
+    @Override
+    public BufferObjectDataInput createObjectDataInput(byte[] data) {
+        return delegate.createObjectDataInput(data);
+    }
+
+    @Override
+    public BufferObjectDataInput createObjectDataInput(Data data) {
+        return delegate.createObjectDataInput(data);
+    }
+
+    @Override
+    public BufferObjectDataOutput createObjectDataOutput(int size) {
+        return delegate.createObjectDataOutput(size);
+    }
+
+    @Override
+    public ManagedContext getManagedContext() {
+        return delegate.getManagedContext();
+    }
+
+    @Override
+    public PortableReader createPortableReader(Data data)
+            throws IOException {
+        return delegate.createPortableReader(data);
+    }
+
+    @Override
+    public PortableContext getPortableContext() {
+        return delegate.getPortableContext();
+    }
+
+    @Override
+    public void disposeData(Data data) {
+        delegate.disposeData(data);
+    }
+
+    @Override
+    public BufferObjectDataOutput createObjectDataOutput() {
+        return delegate.createObjectDataOutput();
+    }
+
+    @Override
+    public ByteOrder getByteOrder() {
+        return delegate.getByteOrder();
+    }
+
+    @Override
+    public byte getVersion() {
+        return delegate.getVersion();
+    }
+
+    @Override
+    public void dispose() {
+        delegate.dispose();
+    }
+
+    private static void addSerializedSample(Object obj, byte[] bytes) {
+        String className = obj.getClass().getName();
+        SERIALIZED_SAMPLES_PER_CLASS_NAME.putIfAbsent(className, new CopyOnWriteArrayList<byte[]>());
+        List<byte[]> samples = SERIALIZED_SAMPLES_PER_CLASS_NAME.get(className);
+        if (samples.size() < MAX_SERIALIZED_SAMPLES_PER_CLASS) {
+            samples.add(bytes);
+        }
+    }
+
+    private static boolean shouldAddSerializedSample(Object obj) {
+        Class klass = obj.getClass();
```

Comment:
Maybe it could be worth to cache a `false` by the class name (in a `HashSet` or so), since once it's `false` it will never be `true` again, if I'm not mistaken.